lib/workflow/postgres/postgres.go

```diff
@@ -0,0 +1,726 @@
+	}
+
+	_, err = pg.db.ExecContext(context.TODO(), `
+INSERT INTO recurring_workflow_schedules(name, workflow_name, parameters, recurrence, enabled, next_run_at)
+VALUES ($1, $2, $3, $4, $5, $6)`,
+		scheduleName, workflowName, parameters, recurrence, enabled, nextRunAt)
+	if err != nil {
+		return wrapErr(err, "could not update workflow schedule")
+	}
+
+	return wrapErr(tx.Commit(), "failed to commit workflow schedule update")
+}
+
+func (pg *PostgresBackend) EnqueueWorkflow(ctx context.Context, w *backend.WorkflowInstance) error {
+	wrapErr := func(err error, msg string) error {
+		return errors.Wrap(err, msg)
+	}
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	tx, err := pg.db.BeginTx(ctx, nil)
+	if err != nil {
+		return wrapErr(err, "failed to begin enqueue workflow transaction")
+	}
+
+	row := tx.QueryRowContext(ctx, enqueueWorkflowQuery, w.InstanceName, w.WorkflowName, w.Parameters)
+	var count sql.NullInt64
+	err = row.Scan(&count)
+	if err != nil {
+		return wrapErr(err, "failed to enqueue workflow")
+	}
+	if err := tx.Commit(); err != nil {
+		return wrapErr(err, "failed to commit enqueue workflow")
+	}
+	if count.Int64 == 0 {
+		return workflow.ErrWorkflowInstanceExists
+	}
+	return nil
+}
+
+func (pg *PostgresBackend) DequeueWorkflow(ctx context.Context, workflowNames []string) (*backend.WorkflowEvent, backend.WorkflowCompleter, error) {
+	ctx, cancel := context.WithCancel(ctx)
+
+	tx, err := pg.db.BeginTx(ctx, nil)
+	if err != nil {
+		cancel()
+		return nil, nil, errors.Wrap(err, "failed to dequeue workflow")
+	}
+
+	stmt, err := tx.Prepare(enqueueTaskQuery)
+	if err != nil {
+		cancel()
+		return nil, nil, err
+	}
+	// TODO: allow multiple workflow names
+	row := tx.QueryRowContext(ctx, dequeueWorkflowQuery, pq.Array(workflowNames))
+	event := &backend.WorkflowEvent{}
+	workc := &PostgresWorkflowCompleter{
+		ctx:             ctx,
+		enqueueTaskStmt: stmt,
+		tx:              tx,
+		cancel:          cancel,
+	}
+	var taskResultID sql.NullInt64
+	err = row.Scan(
+		&workc.wid,
+		&event.Instance.InstanceName,
+		&event.Instance.WorkflowName,
+		&event.Instance.Status,
+		&event.Instance.Parameters,
+		&event.Instance.Payload,
+		&workc.eid,
+		&event.Type,
+		&taskResultID,
+		&event.EnqueuedTaskCount,
+		&event.CompletedTaskCount,
+	)
+	if err == sql.ErrNoRows {
+		tx.Commit() // nolint: errcheck
+		cancel()
+		return nil, nil, workflow.ErrNoWorkflowInstances
+
+	} else if err != nil {
+		cancel()
+		return nil, nil, errors.Wrap(err, "failed to dequeue workflow")
+	}
+
+	event.InstanceID = workc.wid
+
+	if event.Type == backend.TaskComplete {
+		event.CompletedTaskCount++
+		row := tx.QueryRowContext(ctx, getTaskResultQuery, taskResultID)
+		tr := backend.TaskResult{}
+		err = row.Scan(
+			&tr.TaskName, &tr.Parameters, &tr.Status, &tr.ErrorText, &tr.Result)
+		if err != nil {
+			cancel()
+			// FIXME FIXME FIXME
+			// TODO is this an infinite loop (this is panic situation)
```

Comment:
We've convinced ourselves this is actually the right thing to do.  But we should wrap this error and probably log a warning.